File: backend/app/models/transfer.py
from datetime import datetime, timezone

from sqlalchemy import (
    DateTime,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Text,
)
from sqlalchemy.orm import (
    Mapped,
    mapped_column,
    relationship,
)

from app.database.database import Base
from app.models.transfer_status import TransferStatus


class Transfer(Base):
    __tablename__ = "transfers"

    id: Mapped[int] = mapped_column(
        Integer,
        primary_key=True,
        index=True,
    )

    payment_account_id: Mapped[int | None] = mapped_column(
        ForeignKey("payment_accounts.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )

    quote_id: Mapped[str | None] = mapped_column(
        String(36),
        unique=True,
        nullable=True,
        index=True,
    )

    customer_name: Mapped[str | None] = mapped_column(
        String(120),
        nullable=True,
    )

    customer_phone: Mapped[str | None] = mapped_column(
        String(30),
        nullable=True,
        index=True,
    )

    telegram_username: Mapped[str | None] = mapped_column(
        String(64),
        nullable=True,
        index=True,
    )

    rub_amount: Mapped[float] = mapped_column(
        Numeric(12, 2),
        nullable=False,
    )

    egp_amount: Mapped[float] = mapped_column(
        Numeric(12, 2),
        nullable=False,
    )

    exchange_rate: Mapped[float] = mapped_column(
        Numeric(10, 4),
        nullable=False,
    )

    payment_method: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
    )

    status: Mapped[str] = mapped_column(
        String(40),
        nullable=False,
        default=TransferStatus.PENDING_PAYMENT.value,
        index=True,
    )

    payment_confirmed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        index=True,
    )

    receipt_path: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    rejection_reason: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    payment_account = relationship(
        "PaymentAccount",
        back_populates="transfers",
    )

    russian_recipient = relationship(
        "RussianRecipient",
        back_populates="transfer",
        uselist=False,
        cascade="all, delete-orphan",
        passive_deletes=True,
    )

File: frontend/types/transfer.ts
export type TransferStatus =
  | "pending_payment"
  | "payment_proof_uploaded"
  | "payment_confirmed"
  | "waiting_recipient"
  | "ready_to_send"
  | "rub_sent"
  | "completed"
  | "rejected";

export type PaymentMethod =
  | "vodafone_cash"
  | "instapay"
  | string;

export interface Transfer {
  id: number;
  quote_id: string | null;

  rub_amount?: number;
  egp_amount?: number;
  amount_rub?: number;
  amount_egp?: number;

  payment_method?: PaymentMethod;
  status: TransferStatus;

  customer_name?: string | null;
  customer_phone?: string | null;
  telegram_username?: string | null;

  receipt_path?: string | null;
  receipt_url?: string | null;

  rejection_reason?: string | null;

  created_at?: string;
  updated_at?: string | null;
}

export interface TransferStatistics {
  total: number;
  pending: number;
  receiptUploaded: number;
  approved: number;
  completed: number;
  rejected: number;
}